src/backends/interflop-vprec/interflop_vprec.c

```diff
@@ -273,47 +290,97 @@ void _set_vprec_inst_mode(vprec_inst_mode mode) {
 // Round the float with the given precision
 static float _vprec_round_binary32(float a, char is_input, void *context,
                                    int binary32_range, int binary32_precision) {
+  t_context *currentContext = (t_context *)context;
+
+  /* test if a is special cases */
   if (!isfinite(a)) {
     return a;
   }
 
-  /* round to zero or set to infinity if underflow or overflow compare to
+  /* round to zero or set to infinity if underflow or overflow compared to
    * VPRECLIB_BINARY32_RANGE */
   int emax = (1 << (binary32_range - 1)) - 1;
   int emin = (emax > 1) ? 1 - emax : -1;
 
+  /* in absolute error mode, the error threshold also gives the possible
+   * underflow limit */
+  if ((currentContext->relErr == true) && (currentContext->absErr == true)) {
+    if (currentContext->absErr_exp > emin)
+      emin = currentContext->absErr_exp;
+  } else if (currentContext->absErr == true) {
+    emin = currentContext->absErr_exp;
+  }
+
   binary32 aexp = {.f32 = a};
 
   aexp.s32 = ((FLOAT_GET_EXP & aexp.u32) >> FLOAT_PMAN_SIZE) - FLOAT_EXP_COMP;
 
-  /* check for overflow or underflow in target range */
+  /* check for overflow in target range */
   bool sp_case = false;
   if (aexp.s32 > emax) {
     a = a * INFINITY;
     sp_case = true;
   }
 
+  /* check for underflow in target range */
   if (aexp.s32 <= emin) {
-    if ((((t_context *)context)->daz && is_input) ||
-        (((t_context *)context)->ftz && !is_input)) {
+    if ((currentContext->daz && is_input) ||
+        (currentContext->ftz && !is_input)) {
       a = 0;
     } else {
-      a = handle_binary32_denormal(a, emin, aexp.u32, binary32_precision);
+      if ((currentContext->relErr == true) &&
+          (currentContext->absErr == true)) {
+        /* vprec error mode all */
+        if ((-1) * currentContext->absErr_exp < binary32_precision)
+          a = handle_binary32_denormal(a, emin, aexp.u32,
+                                       (-1) * currentContext->absErr_exp);
+        else
+          a = handle_binary32_denormal(a, emin, aexp.u32, binary32_precision);
+      } else if (currentContext->absErr == true) {
+        /* vprec error mode abs */
+        a = handle_binary32_denormal(a, emin, aexp.u32,
+                                     (-1) * currentContext->absErr_exp);
```

Comment:
The -1 here is strange and should be explained.

Reply:
Explanation: on the command line, the magnitude of the absolute error is specified as a power of 2.
In this case, I made the unreasonable assumption that the power of two would be negative, as we're exploring precision...
But handle_binary32_denormal has as fourth parameter the precision of the new format, which is a positive number.

Should be changed to: abs(currentContext->absErr_exp)

Reply:
I've made the changes.